maestro-android/src/androidTest/java/dev/mobile/maestro/MaestroDriverService.kt

```diff
@@ -196,6 +206,66 @@ class Service(
         }
     }
 
+    override fun setLocation(
+        request: MaestroAndroid.SetLocationRequest,
+        responseObserver: StreamObserver<MaestroAndroid.SetLocationResponse>
+    ) {
+        val latitude = request.latitude
+        val longitude = request.longitude
+        val accuracy = 0.0F
+
+        val locMgr = InstrumentationRegistry.getInstrumentation()
+            .context
+            .getSystemService(LOCATION_SERVICE) as LocationManager
+
+        locMgr.addTestProvider(
+            LocationManager.GPS_PROVIDER,
+            "requiresNetwork" === "",
+            "requiresSatellite" === "",
+            "requiresCell" === "",
+            "hasMonetaryCost" === "",
+            "supportsAltitude" === "",
+            "supportsSpeed" === "",
+            "supportsBearing" === "",
+            Criteria.POWER_LOW,
+            Criteria.ACCURACY_FINE
+        )
+
+        val newLocation = Location(LocationManager.GPS_PROVIDER)
+
+        newLocation.latitude = latitude
+        newLocation.longitude = longitude
+        newLocation.accuracy = accuracy
+        newLocation.altitude = 0.0
+        newLocation.accuracy = 500F
+        locMgr.setTestProviderEnabled(LocationManager.GPS_PROVIDER, true)
+
+        val handler = Handler(Looper.getMainLooper())
+
+        fun postLocation() {
+            handler.post {
+                newLocation.setTime(System.currentTimeMillis())
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
+                    newLocation.elapsedRealtimeNanos = SystemClock.elapsedRealtimeNanos()
+                }
+                locMgr.setTestProviderStatus(
+                    LocationManager.GPS_PROVIDER,
+                    LocationProvider.AVAILABLE,
+                    null, System.currentTimeMillis()
+                )
+
+                locMgr.setTestProviderLocation(LocationManager.GPS_PROVIDER, newLocation)
+
+                postLocation()
```

Comment:
This seems like a recursive call, what is the terminal condition of this recursion 🤔 ?

Reply:
The terminal condition is `locationCounter != version`. Note that this is a non-blocking recursion since we are using `Handler` to add message to the end of the message queue

Reply:
Also, worth noting, it is expected that the recursion loop will continue indefinitely to make sure the location remains consistent and is not overwritten by other location providers (as I have observed without the loop)